Refactor base58 decoding (RIPD-1679):

seelabs opened the pull request:
1) Change the interface to the decoding functions so the caller provides the
buffer that is decoded into.

This allows for an implementation that allocates less memory from the heap, and
potentially makes less memory copies. Since all token types except public keys
have know decoded sizes, it is easy for the caller to provide the buffer. If the
decoded result does not fit exactly in the provided buffer (except for public
keys), the decode fails, as the encoded string cannot represent the expected
token. A flag may be passed to the low-level decoding functions to allow to
decode into a smaller buffer. This is used for public keys, which may decode to
either 32 or 33 bytes.

2) Securely zero memory used in decoding in more situations.

When the low-level decoding routines are decoding a potential secret - like
seeds - then intermediate memory is securely zeroed. When decoding non-secrets -
like account ids - then it does not zero the memory.

3) Use a new algorithm for base 58 decoding.

The algorithm first encodes the base 58 number as a base 58^10 number. All these
coefficients will fit into a 64-bit number. This base 58^10 is then decoded
using boost's multi-precision integer library. This greatly improves decoding
times. My benchmark shows approx 10x improvement on 256-bit decodes, not
including the checksum. An end-to-end test of the old and new decoder shows a
2x-4x improvement, depending on what's being decoded.

4) Decode ripple lib encoded seeds at a low level.

ED25519 encoded with ripple lib used a special encoding. Instead of
<TokenId::FamilySeed><16-bytes seed data>, it uses
<TokenId::None><0xE1><0x4B><16-bytes seed data>. This encoding is handled at the
lowest level. A special flag is passed into the decoder to control if this
decoding should be checked for or not. This simplifies calling client code.
Although it does make testing more complex.

Note that I have not finished writing all the tests I intend for this PR. However, it was suggested to get this in review before all the tests were finished. Also note that as of now some of test are marked "manual" b/c they take ~10 minutes to run.

nbougalis commented on src/ripple/basics/Slice.h:
```diff
@@ -143,6 +166,11 @@ class Slice
+/** An immutable linear range of bytes. */
+using Slice = SliceImpl<false>;
+/** A mutable linear range of bytes. */
+using MutableSlice = SliceImpl<true>;
```
Food for thought: the [Networking TS](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/n4734.pdf) defines two very similar types named `mutable_buffer` and `const_buffer`.

seelabs replied:
Thanks. Yeah, we might want to get rid of our slice classes all together and replace them (but obviously, not now).

nbougalis commented on src/ripple/basics/Slice.h:
```diff
@@ -34,34 +37,55 @@
+    /** Can assign from a mutable slice to a non-mutable slice */
+    template<class T,
+             class = std::enable_if_t<!Mutable && std::is_same<T, SliceImpl<true>>::value>>
+    SliceImpl& operator= (/*SliceImpl<true>*/T const& rhs) noexcept
```
Remove the commented-out type.

seelabs replied:
Will do.

seelabs replied:
addressed in e71cfc80b1583e5055c9f4ed5828b2e7c7deeeb4

nbougalis commented on src/ripple/protocol/SecretKey.h:
```diff
@@ -72,25 +75,25 @@ class SecretKey
     begin() const noexcept
     {
-        return buf_;
+        return buf_.data();
```
Why not `buf_.cbegin();`?

seelabs replied:
Will do.

seelabs replied:
Addressed in e71cfc80b1583e5055c9f4ed5828b2e7c7deeeb4

nbougalis commented on src/ripple/protocol/SecretKey.h:
```diff
@@ -72,25 +75,25 @@ class SecretKey
     end() const noexcept
     {
-        return buf_ + sizeof(buf_);
+        return buf_.data() + buf_.size();
```
Why not `buf_.cend();`?

seelabs replied:
Will do.

seelabs replied:
Addressed in e71cfc80b1583e5055c9f4ed5828b2e7c7deeeb4

nbougalis commented on src/ripple/protocol/SecretKey.h:
```diff
@@ -35,8 +35,11 @@ namespace ripple {
+    boost::optional<SecretKey>
+    parseBase58SecretKey (TokenType type, std::string const& s);
 public:
     using const_iterator = std::uint8_t const*;
```
Should probably be `std::array<...>::const_iterator`

seelabs replied:
Will do.

seelabs replied:
Addressed in e71cfc80b1583e5055c9f4ed5828b2e7c7deeeb4

nbougalis commented on src/ripple/protocol/Seed.h:
```diff
@@ -115,8 +116,13 @@ template <>
+
 /** Attempt to parse a string as a seed */
-boost::optional<Seed>
+boost::optional<std::pair<Seed, boost::optional<KeyType>>>
```
Ugh, I'm not sure I like an `optional` that contains a `pair` that contains an `optional`. It's like Russian nesting dolls. I don't have a better alternative to offer. @HowardHinnant, @scottschurr?

seelabs replied:
I considered making `optional<KeyType>` a member of `Seed`, but decided to keep it out of the class (as most of the time it will be empty). I also didn't want another struct of a `Seed + KeyType` as we already have a `Seed`. If there are objections to this interface I'll add it to the `Seed` class. But I'll also wait for other suggestions before I do so.

seelabs replied:
Addressed in a623be578ec6ef192a09cae3d123dcdb19f374d0

Note: I addressed it by adding an `optional<KeyType>` member to seed. As you know, seeds shouldn't have KeyTypes. A single seed should be able to generate multiple keys of multiple types. However, that isn't how it's used. Given the current conflating of seeds and keys I think the cleanest solution is to add the `optional<KeyType` to seed.

seelabs commented:
Note: I see the jenkins failures caused by the clang warning. This warning is a false alarm and has been fixed in more recent versions of clang. I'd like to see if we can update the clang we use on jenkins. If that can't be done, I'll fix the code here to resolve the warning.

nbougalis commented on src/ripple/protocol/SecretKey.h:
```diff
@@ -35,10 +35,13 @@ namespace ripple {
 class SecretKey
```
No change needed, but here's a random thought: I notice that in another part of this commit we derive from `std::array`. _If_ that is OK (@HowardHinnant?) then we may be able to significantly simply `SecretKey`. Given:

    class SecretKey : public std::array<std::uint8_t, 32>

then there's no need to have `data()`, `size()`, `begin()`, `end()`, `cbegin()` and `cend()`. In fact, we could almost get away with:

    using SecretKey = std::array<std::uint8_t, 32>;

HowardHinnant replied:
Normally I wouldn't get too worked up about this.  But with `SecretKey` I think it is worth the effort to keep it its own type.  It would probably never happen, but I would hate for there to be an even remote chance of accidentally sending a `SecretKey` to a function that was expecting a `std::array`.

nbougalis replied:
Agreed. I was mostly curious about the semantics of deriving from `std::array`. Is that legal?

HowardHinnant replied:
It's legal.  But not generally recommended.  I consider it ok if the derived class _is_ an `array`, and _never_ expects to be destructed via a pointer to an `array`.

seelabs commented:
I just pushed three new commits:
1) The first merges the manual/non-manual test classes into a single class. I reduced the number of iterations to a reasonable amount, and I use `unittest-arg` to set a larger value when I want a longer test.
2) I added a minor test-case to make sure boost's mutliprecision library treats leading zeroes when calling `export_bits` as assumed.
3) I was mistaken that public keys could be 33 or 32 bytes. They are always 33 bytes. This allowed me to cleanup the interface to the `decodeBase58` token so it returns a `bool` and removed the `allowResize` parameter.

jwbusch commented on src/ripple/protocol/impl/Seed.cpp:
```diff
@@ -57,10 +60,38 @@ Seed::Seed (uint128 const& seed)
+{
+    using uint256::uint256;
+    zero_after_use_uint256() = default;
+    zero_after_use_uint256(zero_after_use_uint256 const& other) : uint256{other}
```
= default?

seelabs replied:
Addressed in 91c6c712b0973af30980b849d50e33a4f715d449

jwbusch commented on src/ripple/protocol/tokens.h:
```diff
@@ -90,24 +96,83 @@ base58EncodeTokenBitcoin (TokenType type, void const* token, std::size_t size);
+decodeBase58(
+    Slice s,
+    MutableSlice result,
+    bool allowResize);
```
would prefer an enum over a bool parameter. Otherwise you have to add comments like `..., /*allowResize*/ true)` at the callsite. 

vs. something like `enum class decodeBase58Options { None, AllowResize };` then `..., decodeBase58Options::AllowResize);`

No more comment needed, self documenting, easy to add more options (or turn it into a bit field).

seelabs replied:
Addressed in 91c6c712b0973af30980b849d50e33a4f715d449
Note: This was addressed by removing the `allowResize` param altogether. This low-level routine was renamed to `decodeBase58Resizable` and it always allows resizing. It was only called once (apart from testing); Most code uses the higher-leve token decoding functions which never allow resizing.

jwbusch commented on src/ripple/basics/Slice.h:
```diff
@@ -34,34 +37,55 @@
     of the underlying memory.
 */
-class Slice
+template<bool Mutable>
```
Another place I'd prefer an enum... `Slice<true> Slice<false>` vs. `Slice<Immutable> Slice<Mutable>` You have typedefs so it's less of an issue for code readability, but in the debugger you'll see `Slice<true> Slice<false>` and have to remind yourself what those mean. I *think* (at least in MSVC) you'll get readable enum values in the debugger.

seelabs replied:
Addressed in 91c6c712b0973af30980b849d50e33a4f715d449

I made the change, but I'm ambivalent about it. We need to namespace the enum somehow - either with a `enum class` or with a namespace. So it's going to be `SliceImpl<true>` vs. `SliceImpl<SliceImplMutability::yes>`. If you use an `enum class`, the parameter will not be convertable to a bool. This clutters up the `enable_if` code. Before:
```
        class = std::enable_if_t<
            !Mutable &&
            std::is_same<T, SliceImpl<true>>::value>>
```
After
```
        class = std::enable_if_t<
            Mutable == SliceImplMutability::no &&
            std::is_same<T, SliceImpl<SliceImplMutability::yes>>::value>>
```
We can get around the conversion to bool by defining a `constexpr` class member, so that helps (and I did so). The `no/yes` enum is fine for this (this class is likely to go away), though I'm not convinced we want to do that all the time for boolean template parameters.

BTW, I didn't do `mutable`, `immutable` as the types because `mutable` is a keyword and our enums are lowercase.

jwbusch replied:
You can always mark my comments "won't fix" if you don't think it's useful/too much work for too little gain :) bool input params for templates + functions are just a pet-peeve of mine. I've just seen too many bugs from it and wasted too much time debugging/reading code with them. 

I'm almost thinking there could be a pattern of giving names to these kind of bools and keeping them generic (not specific to one class). `enum class IsMutable{ no, yes }; enum class IsAdmin { no, yes }; enum class IsHumanReadable{ no, yes };` etc. Not a perfect solution, but maybe slightly better?

jwbusch commented on src/ripple/protocol/impl/tokens.cpp:
```diff
@@ -255,28 +490,75 @@ decodeBase58 (std::string const& s,
+    ExtraB58Encoding extraB58Encoding = ExtraB58Encoding::None;
+
+    if (type == TokenType::FamilySeed &&
+        metadata.encodingType[0] == std::uint8_t(0xE1) &&
```
this is duplicated in isRippleLibEncoded(), can it be shared somewhere? i.e. I don't where these constants came from, but at least don't duplicate them :)

jwbusch replied:
Oh, you could just call the function on metaData, it was the same class :) `metaData.isRippleLibEncoded()`

seelabs replied:
Addressed in 91c6c712b0973af30980b849d50e33a4f715d449

jwbusch commented on src/ripple/protocol/PublicKey.h:
```diff
@@ -58,12 +58,14 @@ namespace ripple {
-    std::uint8_t buf_[33]; // should be large enough
+private:
+    bool empty_{true};
+    std::array<std::uint8_t, 33> buf_; // all supported public keys are 33 bytes
```
although the empty is just used to save from needlessly copying data, the iterator functions would stop working because begin wouldn't be accessible, so nevermind :)

seelabs replied:
I like the idea of using `optional`, but yeah, it complicates the iterators. Let's keep the bool member for now.

jwbusch commented on src/ripple/protocol/impl/tokens.cpp:
```diff
@@ -77,12 +88,16 @@ digest2 (Args const&... args)
+checksum(void* out, void const* message, std::size_t size)
+{
+    auto const h = digest2<sha256_hasher>(message, size);
+    std::memcpy(out, h.data(), 4);
```
Where does 4 come from? -> put in a constexpr?

seelabs replied:
Addressed in 91c6c712b0973af30980b849d50e33a4f715d449

jwbusch commented on src/ripple/protocol/impl/tokens.cpp:
```diff
@@ -162,90 +180,307 @@ encodeToken (TokenType type,
+        std::uint8_t const* dataStart = &tmp[1];
+        std::uint8_t const* dataEnd = &tmp[0] + numWritten - 4;
+        if (flags.test(DecodeBase58Detail::maybeRippleLibEncoded) &&
+            static_cast<TokenType>(metadata.tokenType) == TokenType::None &&
```
I see those same constants again... 0xE1 and 0x4B... along with lots of other magic numbers. `dataEnd - dataStart == 18`?

seelabs replied:
Addressed in 91c6c712b0973af30980b849d50e33a4f715d449

seelabs commented on src/ripple/protocol/SecretKey.h:
```diff
@@ -35,10 +35,13 @@ namespace ripple {
+    parseBase58<SecretKey> (TokenType type, std::string const& s);
 public:
-    using const_iterator = std::uint8_t const*;
+    using const_iterator = std::array<std::uint8_t, 32>::const_iterator;
```
Addressed in 91c6c712b0973af30980b849d50e33a4f715d449

seelabs commented:
Pushed a commit to: 1) remove the `allowResize` flag entirely; 2) replace some boos with no/yes enums; 3) Define symbolic constants for some magic numbers in tokens.cpp

mellery451 commented on src/ripple/protocol/impl/tokens.cpp:
```diff
@@ -149,103 +174,324 @@ encodeBase58(
+                                                      38068692544,
+                                                      2207984167552,
+                                                      128063081718016,
+                                                      7427658739644928};
```
no change needed, but I wonder if something like:
```
    constexpr std::array<std::uint64_t, 10> pow58{
        1ull,
        58ull,
        58ull*58,
        58ull*58*58,
        58ull*58*58*58,
        58ull*58*58*58*58,
        58ull*58*58*58*58*58,
        58ull*58*58*58*58*58*58,
        58ull*58*58*58*58*58*58*58,
        58ull*58*58*58*58*58*58*58*58};
```
would be easier for others to read. Either way, the values are clear, even if you can't do 58^9 in your head.

nbougalis replied:
Or, we can just write some code and let the compiler do all the work:

```
constexpr auto pow58 = []()
{
    std::array<std::uint64_t, 10> powtab { 1 };

    for (std::size_t i = 1; i != powtab.size(); ++i)
        powtab[i] = 58 * powtab[i - 1];

    return powtab;
}();
```

seelabs replied:
Lambda aren't constexpr until c++-17, so we can't use that yet.

nbougalis replied:
Oh, duh.

nbougalis commented on src/ripple/basics/Slice.h:
```diff
@@ -34,34 +37,64 @@
     of the underlying memory.
 */
-class Slice
+template<SliceImplMutability M>
```
This `SliceImplMutability` business seems like too much effort and a bit too convoluted. Why not have something like this:

    template <typename T,
        typename = std::enable_if_t<std::is_same<std::remove_cv_t<T>, std::uint8_t>::value>
    >
    class SliceImpl
    {
        T* data_ = nullptr;
    };

    /** An immutable linear range of bytes. */
    using Slice = SliceImpl<std::uint8_t const>;
    /** A mutable linear range of bytes. */
    using MutableSlice = SliceImpl<std::uint8_t>;

seelabs replied:
Addressed in bcaaf38a3e0e9516fd5ac28d73a7721de73188c0

BTW, I changed the `enable_if` into a `static_assert`. I'd use `enable_if` for an overload-set. For an assumption that I want to error out I'd rather use `static_assert`.

nbougalis commented on src/ripple/protocol/impl/PublicKey.cpp:
```diff
@@ -179,23 +181,25 @@ ed25519Canonical (Slice const& sig)
 PublicKey::PublicKey (Slice const& slice)
 {
-    if(! publicKeyType(slice))
+    if (slice.size() != buf_.size() || !publicKeyType(slice))
```
The size check here is redundant: `publicKeyType` does it for you and will return an unseated `boost::optional` if the size isn't right (i.e. exactly 33 bytes).

seelabs replied:
Addressed in bcaaf38a3e0e9516fd5ac28d73a7721de73188c0

Note: I understood that the check was redundant. I made the change because I had initially thought a key could 32 or 33 bytes and if I saw code like this in the constructor it would have helped me. However, given the change removing the `size_` (and as you point out the code in `publicKeyType`) it very clear the key is always 33 bytes. Removed the check.

nbougalis commented on src/ripple/protocol/impl/SecretKey.cpp:
```diff
@@ -33,19 +33,19 @@ namespace ripple {
 SecretKey::SecretKey (std::array<std::uint8_t, 32> const& key)
 {
-    std::memcpy(buf_, key.data(), key.size());
+    std::memcpy(buf_.data(), key.data(), key.size());
```
Why not simply:

    SecretKey::SecretKey (std::array<std::uint8_t, 32> const& key)
        : buf_(key)
    {
    }

seelabs replied:
Addressed in 4ecdd58c54c23ffbb297be83b8252e53631287fe

nbougalis commented on src/ripple/protocol/impl/tokens.cpp:
```diff
@@ -77,13 +97,17 @@ digest2 (Args const&... args)
     @note This checksum algorithm is part of the client API
 */
 void
```
The `checksum` functions should probably be marked `static`

seelabs replied:
Addressed in 4ecdd58c54c23ffbb297be83b8252e53631287fe

nbougalis commented on src/ripple/rpc/impl/RPCHelpers.cpp:
```diff
@@ -671,44 +652,71 @@ keypairForSignature(Json::Value const& params, Json::Value& error)
-        seed = RPC::parseRippleLibSeed(params[secretType]);
+        // Check the key type and set the error if needed. returns true if no error,
+        // false if error
+        auto checkKeyType = [&error](boost::optional<KeyType> const& expected, boost::optional<KeyType> const& got) -> bool{
```
Two comments:

1) Is a trailing return required here?
2) I'm wondering whether an almost 50-line lambda deserves to be a standlone function.

seelabs replied:
Addressed in f69d726530ea5dc51951985db5b2a88ea8c9252e

BTW, it was only ~20 lines, not 50 :-)

seelabs commented on src/ripple/protocol/PublicKey.h:
```diff
@@ -80,49 +82,49 @@ class PublicKey
     slice() const noexcept
     {
-        return { buf_, size_ };
+        return {buf_.data(), buf_.size()};
```
Addressed in 4ecdd58c54c23ffbb297be83b8252e53631287fe

Thanks for spotting that. Yes, I missed that and it was a bug.

seelabs commented on src/ripple/protocol/Seed.h:
```diff
@@ -22,18 +22,19 @@
 namespace ripple {
 
-/** Seeds are used to generate deterministic secret keys. */
+/** Seeds are used to a generate deterministic secret key. */
```
Address in 4ecdd58c54c23ffbb297be83b8252e53631287fe

seelabs commented:
Pushed a commit to address some of @nbougalis comments.

seelabs commented on src/ripple/protocol/impl/PublicKey.cpp:
```diff
@@ -39,11 +39,13 @@ template<>
-        return boost::none;
-    return PublicKey(pks);
+    PublicKey result;
+    result.empty_ = false;
```
The root of the problem is `PublicKey` is default constructable, and I don't think it should be. Anything using a default constructed public key should be changed to use a `boost::optional<PublicKey>` and any constructed `PublicKey` should be valid. This removes the `empty_` junk from public key and makes explicit the assumptions about how the key is used at the site.

I spent about an hour making this change to see what it looked like, and I gave up. There are way too many places using default constructed keys now to make the change in this PR. However, that would be my solution to fixing this interface.

sublimator commented:
/**
  * @param {String} desiredPrefix - desired prefix when base58 encoded with
  *                                 checksum
  * @param {Number} payloadLength - number of bytes encoded not incl checksum
  * @return {Array} version
  */
  findPrefix(desiredPrefix, payloadLength) {
    if (this.base !== 58) {
      throw new Error('Only works for base58');
    }
    const totalLength = payloadLength + 4; // for checksum
    const chars = (Math.log(Math.pow(256, totalLength)) / Math.log(this.base));
     // (x, x.8] -> x+1, (x.8, x+1) -> x+2
    const requiredChars = Math.ceil(chars + 0.2);
    const padding = this.alphabet[Math.floor((this.alphabet.length) / 2) - 1];
    const template = desiredPrefix + new Array(requiredChars + 1).join(padding);
    const bytes = this.decodeRaw(template);
    const version = bytes.slice(0, -totalLength);
    return version;
  }


Hrmmmmmm

nbougalis commented:
@sublimator "Hrmmmmmm" is a little confusing. Is there something specific that's of concern to you?

sublimator commented:
haha all the things
